### PR TITLE
[exporterhelper] Add WithRequestQueue option to the exporter

### PR DESCRIPTION
**Pull request by dmitryax**
Introduce a way to enable queue in the new exporter helper with a developer interface suggested in https://github.com/open-telemetry/opentelemetry-collector/pull/8248#discussion_r1302102261. 

The new configuration interface for the end users provides a new `queue_size_items` option to limit the queue by a number of spans, log records, or metric data points. The previous way to limit the queue by number of requests is preserved under the same field, `queue_size,` which will later be deprecated through a longer transition process.

Tracking issue: https://github.com/open-telemetry/opentelemetry-collector/issues/8122

### CODE REVIEW

File: exporter/exporterhelper/internal/request/request.go
Comment (dmitryax) [inline exporter/exporterhelper/internal/request/request.go:54]: If we expose the queue interface, this probably can be exposed as an optional request interface as well. Then the queue `Produce` method can accept a public `Request` instead of the internal one

File: exporter/exporterhelper/internal/request/request.go
Comment (dmitryax) [inline exporter/exporterhelper/internal/request/request.go:61]: If we expose the queue interface now, this can be exposed as an optional Request interface as well. Then the `Queue.Produce` method can accept the public Request type instead of the internal one

Comment (dmitryax) [reply exporter/exporterhelper/internal/request/request.go]: Not relevant anymore after recent changes

File: exporter/exporterhelper/internal/queue_capacity.go
Comment (splunkericl) [inline exporter/exporterhelper/internal/queue_capacity.go:68]: This can overflow from 0 to 4294967295. 

I tested the new changes with persistent queue and saw that:
- when persistent queue initialize, it loads some `currentlyDispatchedItems` and consume it. Once an item is consumed, this number becomes overflow and all future `Claim` becomes blocked.

Would it make sense to safe guard it from overflowing(always above 0)? Otherwise it is hard to ensure callers to not accidentally call `Release`.

Comment (dmitryax) [reply exporter/exporterhelper/internal/queue_capacity.go:68]: Thanks for the reports! It's a bug indeed. Let me fix

Comment (dmitryax) [reply exporter/exporterhelper/internal/queue_capacity.go]: I implemented the more resilient queue size tracking for the persistent queue

File: exporter/exporterhelper/queue.go
Comment (bogdandrutu) [inline exporter/exporterhelper/queue.go:97]: Provide a "PersistentQueueSettings" to allow extending. Include the Marshaler/Unmarshaler for the moment.

File: exporter/exporterhelper/queue.go
Comment (bogdandrutu) [inline exporter/exporterhelper/queue.go:29]: Should we offer 2 sizes or a size + enum saying count num requests or count num items eventually count bytes if we want?

File: exporter/exporterhelper/queue.go
Comment (bogdandrutu) [inline exporter/exporterhelper/queue.go:17]: Let's probably start there?

File: exporter/exporterhelper/queue.go
Comment (dmitryax) [reply exporter/exporterhelper/queue.go:29]: I was thinking about it. From the user's perspective, specifying the enum field might be less obvious. Also, setting `queue_element: item` (assuming we find a better field name) where the item represents a span or metric data point seems more confusing than having a set of predefined documented fields. Let me know if you have any other specific suggestions. For now, I think `queue_size`, `queue_size_items`, and potentially `queue_size_mib` is an easier to use interface.

Comment (bogdandrutu) [reply exporter/exporterhelper/queue.go:29]: maybe use suffix? `queue_size: 10[B|KB|MB..]` (applies to size in mb) then `queue_size: 10[requests]?` (requests) and `queue_size: 10items`

File: exporter/exporterhelper/queue.go
Comment (dmitryax) [reply exporter/exporterhelper/queue.go]: Makes sense. Applied

File: exporter/exporterhelper/queue.go
Comment (dmitryax) [reply exporter/exporterhelper/queue.go]: Ok. Created the `config/configqueue` module

File: exporter/exporterhelper/queue.go
Comment (dmitryax) [reply exporter/exporterhelper/queue.go]: I'm still hesitant that it'll make the user experience easier with `items` being ambiguous.

Having it as a part of the field names gives some flexibility to the component implementations, like they can have different name for the `items` field where it's not directly mapped to spans, datapoint log records. Having the parsing logic hardcoded wouldn't allow that. Also, I can imagine a use case when only one type would be supported by a particular exporter. With different fields, it's possible to restrict the options on the exporter side.

Comment (dmitryax) [reply exporter/exporterhelper/queue.go]: If we agree to keep different fields, maybe we can also rename them by removing `queue_` prefix since it's supposed be under `sending_queue` group anyway. Can be 
- `max_size_items` and `max_size_mib`
- `size_limit_items` and `size_limit_mib`
- `items_limit` and `mib_limit`

File: config/configqueue/queue.go
Comment (bogdandrutu) [inline config/configqueue/queue.go:64]: Would be interested to see how people decide to use this vs QueueConfig in their Configs. When do I pick one or the other?

File: config/configqueue/queue.go
Comment (bogdandrutu) [inline config/configqueue/queue.go:16]: second thoughts: for retry we put it under config because receivers/connectors may also retry adding to a pipeline. For queue do we see an use case to be used outside exporter? If not, consider to have a `exporter/exporterqueue` where we declare the config and the Queue interface?

File: consumer/go.mod
Comment (bogdandrutu) [inline consumer/go.mod:57]: How does this gets here? Seems that we have some sort of circular dependency via the `go.opentelemetry.io/collector` module.

File: exporter/debugexporter/go.mod
Comment (bogdandrutu) [inline exporter/debugexporter/go.mod:39]: Interesting change... Do you know why? This is just educational.

File: exporter/exporterhelper/internal/bounded_memory_queue.go
Comment (bogdandrutu) [inline exporter/exporterhelper/internal/bounded_memory_queue.go:25]: The CreateSettings change can be extracted in a separate PR since we need that anyway.

File: exporter/exporterhelper/internal/persistent_queue.go
Comment (bogdandrutu) [inline exporter/exporterhelper/internal/persistent_queue.go:90]: Why not having a PersistentQueueCreateSettings? Also, to be honest not sure if we should use `Create` here, we use in the Factory case since that is a `CreateProcessor` function, but in case of the `service` objects I think we don't use.

File: exporter/exporterhelper/common.go
Comment (bogdandrutu) [inline exporter/exporterhelper/common.go:136]: Instead of having a new public `QueueFactory` what about we change this to `func WithRequestQueue(queue Queue, numConsumers int)`

File: config/configqueue/queue.go
Comment (dmitryax) [reply config/configqueue/queue.go:64]: `QueueConfig` is used only for exporters that don't support persistent queue. And `PersistentQueueConfig` for exporters that support both. Maybe we don't need to support memory-queue-only exporters. My idea was that some exporters don't want to provide serializers but still want to use custom requests.

File: config/configqueue/queue.go
Comment (dmitryax) [reply config/configqueue/queue.go:16]: Sounds good to me. I don't think there is any use case outside exporters

File: consumer/go.mod
Comment (dmitryax) [reply consumer/go.mod:57]: It comes from `obsreport/obsreporttest` which is not a separate module. We need to get rid of that package

File: exporter/debugexporter/go.mod
Comment (dmitryax) [reply exporter/debugexporter/go.mod:39]: I believe it should be part of https://github.com/open-telemetry/opentelemetry-collector/pull/9178. Not sure why it wasn't caught there

File: exporter/exporterhelper/common.go
Comment (dmitryax) [reply exporter/exporterhelper/common.go:136]: Not sure how to pass `QueueCreateSettings` then. Should we move it to the Start method?

Comment (bogdandrutu) [reply exporter/exporterhelper/common.go:136]: `QueueCreateSettings` - What is in there that user doesn't have access to or you think is "duplicate"?

File: config/configqueue/queue.go
Comment (bogdandrutu) [reply config/configqueue/queue.go:64]: >  My idea was that some exporters don't want to provide serializers but still want to use custom requests.

We can have some default serializers using pdata maybe in that case? May complicate too much the implementation design...

Comment (dmitryax) [reply config/configqueue/queue.go:64]: That would require asking user for Request->pdata translator. Right now we ask for pdata->Request only

File: exporter/exporterhelper/common.go
Comment (dmitryax) [reply exporter/exporterhelper/common.go:136]: Yes, we would need to ask the user to pass at least `component.DataType` and `exporter.CreateSettings` to the persistent queue constructor in addition to the exporter helper itself

File: exporter/exporterhelper/internal/bounded_memory_queue.go
Comment (dmitryax) [reply exporter/exporterhelper/internal/bounded_memory_queue.go:25]: Submitted https://github.com/open-telemetry/opentelemetry-collector/pull/9197

File: exporter/exporterhelper/internal/persistent_queue.go
Comment (dmitryax) [reply exporter/exporterhelper/internal/persistent_queue.go:90]: Makes sense. I renamed it to `QueueSettings/PersistentQueueSettings` in https://github.com/open-telemetry/opentelemetry-collector/pull/9197

File: config/configqueue/queue.go
Comment (dmitryax) [reply config/configqueue/queue.go]: I moved the new API to `exporter/exporterqueue`. Given that both `exporter/exporterhelper` and `exporter/exporterqueue` depend on `exporter/exporterqueue/internal`, I moved it as is to `exporter/internal` for now to keep the changeset simple. Once this PR is merged, I'll distribute that code.

File: config/configqueue/queue.go
Comment (dmitryax) [reply config/configqueue/queue.go]: Otherwise, we can apply the conversion after the queue. This can significantly complicate the implementation, as you mentioned.

File: exporter/exporterhelper/queue.go
Comment (dmitryax) [reply exporter/exporterhelper/queue.go]: @open-telemetry/collector-approvers, can you please share your thoughts on the user interface? The ask is to provide a way to configure queue size in different measures:
1. Number of basic items: log records, spans, data points - this should be the default measure going forward to be consistent with config for batching
2. Number of incoming requests - this is the current measure set in `queue_size`. We likely should keep it as an option even though it's not required.
3. Size in MiB - will be nice to have in the future.

**Options:**

A. Provide different fields for each option
- `max_size_items`, `max_size_requests` and `max_size_mib`
- `size_limit_items`, `size_limit_requests` and `size_limit_mib`
- `items_limit`, `requests_limit` and `mib_limit`

B. Keep `queue_size` and introduce another field for the `queue_measure` enum which can be `items`, `requests` or `MiB`.

C. Support suffixes like `10requests`, `10items` `10[B|KB|MB..]` and keep the `queue_size` field. 

Both _B_ and _C_ would require a plan on how to migrate from the default measure from `requests` to `items`.

Feel free to suggest other ideas or names for the fields.

File: exporter/exporterqueue/queue.go
Comment (bogdandrutu) [inline exporter/exporterqueue/queue.go:45]: Extract "Marshaler" as a type and eventually deprecate `RequestMarshaler` in the exporterhelper?

File: exporter/internal/err.go
Comment (bogdandrutu) [inline exporter/internal/err.go:4]: put it in a `internal/queue` package please.

File: exporter/exporterqueue/queue.go
Comment (bogdandrutu) [inline exporter/exporterqueue/queue.go:20]: Anti pattern to expose internal types like this. Consider to move the definition of Sizer at least.

Also, is the sizer/type not defined by the config arguments? Why do you need to pass it here?

File: exporter/exporterqueue/queue.go
Comment (bogdandrutu) [inline exporter/exporterqueue/queue.go:32]: Should accept Context, Settings and Config. See pattern here https://github.com/open-telemetry/opentelemetry-collector/blob/main/service/extensions/extensions.go#L174

Where Settings are things like TelemetrySettings, etc. In your case Sizer and Capacity should be passed and calculated from Config.

File: exporter/exporterhelper/queue.go
Comment (bogdandrutu) [reply exporter/exporterhelper/queue.go]: Personally I like in this order (from most liked to least like: C > B > A.

> Both B and C would require a plan on how to migrate from the default measure from requests to items without affecting configs applied by users.

Why would you want to change the default measure?

Comment (dmitryax) [reply exporter/exporterhelper/queue.go]: > Main reasons (though we can also validate to not set multiple values) it is confusing if users configure multiple sizes which one to be used, and trying to implement multiple sizes is out of the scope and an overkill.

Right, I wanted to validate that only one field is specified.

> Why would you want to change the default measure?

Because the batch processor is configured in items, I think it should be the same default for the batcher sender. So, it may be confusing to have different measures by default in the ` sending_queue` and the `batching` options.

File: exporter/exporterqueue/queue.go
Comment (dmitryax) [reply exporter/exporterqueue/queue.go]: Defined Marshaler in this package and deprecated `RequestMarshaler`

File: exporter/internal/err.go
Comment (dmitryax) [reply exporter/internal/err.go]: done

File: exporter/exporterqueue/queue.go
Comment (dmitryax) [reply exporter/exporterqueue/queue.go]: done

File: exporter/exporterqueue/queue.go
Comment (dmitryax) [reply exporter/exporterqueue/queue.go]: Removed this field 👍

File: exporter/exporterhelper/queue.go
Comment (dmitryax) [reply exporter/exporterhelper/queue.go]: I removed the `queue_size_items` and left only the backward compatible `queue_size` until we agree on the user config and a migration path. I'll create another issue to move this discussion

**Review by bogdandrutu — APPROVED** [review]

File: exporter/exporterhelper/queue.go
Comment (dmitryax) [reply exporter/exporterhelper/queue.go]: Submitted https://github.com/open-telemetry/opentelemetry-collector/issues/9462

**Pull request merged**
